antipackage.py

```diff
@@ -1,130 +1,401 @@
+            _conv[key] = value
+        else:
+            raise ValueError("'.' is a reserved character")
+    else:
+        del _conv[key]
+
+def repr_conv(chars):
+    """Applies replacements specified by `replace_char`"""
+    for key in _conv:
+        chars = chars.replace(key, _conv[key])
+    return chars
+
+def display_download(count, block, size):
+    """Displays download progress for sizable files
+
+    Notes
+    -----
+    Supply as the 'reporthook' for urlretrieve"""
+    fraction = size/block
+    if fraction>10:
+        percent = int(count*100/fraction)
+        sys.stdout.write("\rDownload in progress... %d%%" % percent)
+        sys.stdout.flush()
+
+# - - - - - - - - - - - - - - - -
+# Setup Base Dir and Pinning File
+# - - - - - - - - - - - - - - - -
+
+def _setup(directory, fname):
+    base_path = os.path.expanduser('~/'+directory)
+    if not os.path.exists(base_path):
+        os.makedirs(base_path)
+    data_path = base_path+'/'+fname
+    if not os.path.isfile(data_path):
+        with open(data_path, 'w') as f:
+            json.dump({},f)
+    return base_path, data_path
+
+_BASE_DIR, _DATA_PATH = _setup('.antipackage','pinnings.json')
+
+# - - - - - - - - - - - - - - - - - - - - -
+# Main Data Management Object and Functions
+# - - - - - - - - - - - - - - - - - - - - -
+
+def pin(path, **kwargs):
+    """Associate a repo with a particular version
+
+    GitHub Paramters
+    ----------------
+    path : str
+        A string indicating a data location via keys delimited
+        by a series of '/' characters. This is an example of a
+        GitHub path - 'github/username/repo'.
+    **kwargs : dict
+        There should only be one keyword argument. That being
+        for either 'branch', 'tag', or 'sha' corrisponding to
+        the kinds of pins associated with a repository.
+        Marking a repo with a branch pin will cause antipackage
+        to pull from the most recent version found on that
+        branch when importing. However marking a repo with a
+        sha or tag pin will force antipackage to draw on
+        version of the repository which orrisponds to that
+        particular commit during all future imports. Default
+        points to 'master' branch of the repository indicated
+        by path.
+    """
+    Author(path,True).write(**kwargs)
+
+def data(path=None):
+    """Get a dictionary of pinning data at the given path
+
+    GitHub Parameters
+    -----------------
+    path : str
+        A string indicating a data location via keys delimited
+        by a series of '/' characters.
+    """
+    with open(_DATA_PATH,'r') as f:
+        if not path:
+            return json.load(f)
+        else:
+            pathlist = path.strip('/').split('/')
+            data = json.load(f)
+            for k in pathlist:
+                data = data.get(k,dict())
+                if not data:
+                    return dict()
+            return data
+
+class Author(object):
```

Comment:
I would rename this class to something like `PackageState` that reflects that this tracks the state of an installed package.